[quant] Add saturate_to_fp16 op for FP16 quant support

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42348 [quant] Use PlaceholderObserver instead of Fp16Observer and NoopObserver
* #42222 [quant][graph] Add support for FP16 dynamic quant
* #42221 [quant] Add FP16Observer for fp16 quant support
* **#42147 [quant] Add saturate_to_fp16 op for FP16 quant support**

Summary:
Op to check the range of a tensor and clamp the values to fp16 range
This operator will be inserted into the graph in subsequent diffs.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_fp16_saturate_op

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22849221](https://our.internmc.facebook.com/intern/diff/D22849221)

supriyar committed:
[quant] Add saturate_to_fp16 op for FP16 quant support
Summary:
Op to check the range of a tensor and clamp the values to fp16 range
This operator will be inserted into the graph in subsequent diffs.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_fp16_saturate_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on aten/src/ATen/native/quantized/cpu/quant_utils.h:
```diff
@@ -136,4 +165,20 @@ static torch::List<int64_t> MakeArgForConv1d(const torch::List<int64_t>& arg,
+  const float kFp16Max = RawUint16ToFp16(0x7BFF);
+  bool found_out_of_range = false;
+  for (int64_t i = 0; i < N; ++i) {
+    if (CheckAndSaturate<float>(kFp16Max, weight + i)) {
```
optional nit: calling a function with side effects inside the if condition is usually not intuitive.  Maybe we can move the calculation into a bool before the if block?

vkuzo commented on aten/src/ATen/native/quantized/cpu/quant_utils.h:
```diff
@@ -5,6 +5,35 @@
+}
+
+template <typename T>
+bool CheckAndSaturate(T max_val, T* element) {
```
optional: I realize this is just moving code around, but this API seems to require careful reading to understand it correctly.  Might be worth a refactor in the future to make it a void return type, and have the status be a var passed by reference.

supriyar committed:
Update on "[quant] Add saturate_to_fp16 op for FP16 quant support"


Summary:
Op to check the range of a tensor and clamp the values to fp16 range
This operator will be inserted into the graph in subsequent diffs.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_fp16_saturate_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add saturate_to_fp16 op for FP16 quant support"


Summary:
Op to check the range of a tensor and clamp the values to fp16 range
This operator will be inserted into the graph in subsequent diffs.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_fp16_saturate_op

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant] Add saturate_to_fp16 op for FP16 quant support"


Summary:
Op to check the range of a tensor and clamp the values to fp16 range
This operator will be inserted into the graph in subsequent diffs.

Test Plan:
python test/test_quantization.py TestQuantizedTensor.test_fp16_saturate_op

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22849221](https://our.internmc.facebook.com/intern/diff/D22849221)

[ghstack-poisoned]
